packages/synapse-sdk/src/storage/manager.ts

```diff
@@ -146,26 +159,55 @@ export class StorageManager {
       }
     }
 
-    // Get the context to use
+    // Get the context(s) to use
     const contexts = options?.contexts ?? [options?.context ?? (await this.createContext(options))]
-    // Convert once
-    const dataBytes = data instanceof ArrayBuffer ? new Uint8Array(data) : data
-    // Calculate pieceCid once
-    const pieceCid = Piece.calculate(dataBytes)
-
-    // Upload using the contexts with piece metadata
-    return await Promise.allSettled(
-      contexts.map((context) =>
-        context.upload(
-          dataBytes,
-          {
+
+    // Multi-context upload handling
+    if (contexts.length > 1) {
+      // Multi-context uploads require Uint8Array to calculate pieceCid once
+      if (!(data instanceof Uint8Array)) {
+        throw createError(
+          'StorageManager',
+          'upload',
+          'Multi-context uploads currently only support Uint8Array. ' +
+            'For streaming uploads to multiple providers, convert your stream to Uint8Array first.'
+        )
+      }
+
+      // Calculate pieceCid once for all contexts
+      const pieceCid = Piece.calculate(data)
+
+      // Upload to all contexts with the same pieceCid
+      return await Promise.allSettled(
+        contexts.map((context) =>
+          context.upload(data, {
             ...options?.callbacks,
             metadata: options?.metadata,
-          },
-          pieceCid
+            pieceCid,
+            signal: options?.signal,
+          })
         )
       )
-    )
+    } else {
+      // Single context upload - supports all data types
+      const context = contexts[0]
+
+      // For Uint8Array, calculate pieceCid once
+      let pieceCid: Piece.PieceCID | undefined
+      if (data instanceof Uint8Array) {
+        pieceCid = Piece.calculate(data)
+      }
+
+      // Upload to single context
+      const result = await context.upload(data, {
+        ...options?.callbacks,
+        metadata: options?.metadata,
+        pieceCid,
+        signal: options?.signal,
+      })
+
+      return [{ status: 'fulfilled' as const, value: result }]
```

Comment:
not necessarily an issue for this PR, but:

I know I commented about us returning `promise.allSettled` instead of `promise.all` previously, but idk how I feel about an SDK returning `Promise<PromiseSettledResult<UploadResult>[]>`.

definitely something consumers should be able to handle, and keeps it flexible without maintainer burden on handling all the edge-cases, but that means consumers will need to, right?

I don't think it's a bad call, just something I think should have an explicit agreement from the core maintainers upon, and maybe some callouts for consumers.